From c4080bc89f2b63feac3ce5eaec20fb519a79fa66 Mon Sep 17 00:00:00 2001
From: Igor Khanin <igor@khanin.biz>
Date: Sat, 04 Oct 2025 16:14:59 +0300
Subject: [PATCH] Fix getting color scheme in x-d-p platform theme

Commit 4ac89da broke reading system color scheme support from the
settings portal of xdg-desktop-portal. The cause is that the metatypes
provided for the parameter and result of the ReadAll method are incorrect and don't match the declared interface which is:

ReadAll (
  IN namespaces as,
  OUT value a{sa{sv}}
)

This patch ensures that the parameter and return types used with QDBus
are correct, and adds warning logs in case of future failures.

Change-Id: I6ac51dbfa1e9f1ebcb59e8398b6c5119da79d1f8
Reviewed-by: Oliver Eftevaag <oliver.eftevaag@qt.io>
(cherry picked from commit 9850c465c900bfc0b43c4ab21e3c361aa39ff417)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---

diff --git a/src/dbus/qdbusmetatype.cpp b/src/dbus/qdbusmetatype.cpp
index 5a86252..033668a 100644
--- a/src/dbus/qdbusmetatype.cpp
+++ b/src/dbus/qdbusmetatype.cpp
@@ -88,6 +88,7 @@
         qDBusRegisterMetaType<QList<qlonglong> >();
         qDBusRegisterMetaType<QList<qulonglong> >();
         qDBusRegisterMetaType<QList<double> >();
+        qDBusRegisterMetaType<QMap<QString, QVariantMap>>();
 
         // plus lists of our own types
         qDBusRegisterMetaType<QList<QDBusVariant> >();
diff --git a/src/plugins/platformthemes/xdgdesktopportal/qxdgdesktopportaltheme.cpp b/src/plugins/platformthemes/xdgdesktopportal/qxdgdesktopportaltheme.cpp
index e4c9b52..e0c0071 100644
--- a/src/plugins/platformthemes/xdgdesktopportal/qxdgdesktopportaltheme.cpp
+++ b/src/plugins/platformthemes/xdgdesktopportal/qxdgdesktopportaltheme.cpp
@@ -126,6 +126,9 @@
         QDBusPendingReply<QVariant> reply = *watcher;
         if (reply.isValid()) {
             d->fileChooserPortalVersion = reply.value().toUInt();
+        } else {
+            qWarning() << "Call for getting org.freedesktop.portal.FileChooser version failed"
+                       << reply.error();
         }
         watcher->deleteLater();
     });
@@ -135,17 +138,21 @@
                                              "/org/freedesktop/portal/desktop"_L1,
                                              "org.freedesktop.portal.Settings"_L1,
                                              "ReadAll"_L1);
-    message << appearanceInterface;
+
+    QStringList namespaces = { appearanceInterface };
+    message << namespaces;
 
     // this must not be asyncCall() because we have to set appearance now
-    QDBusReply<QVariant> reply = QDBusConnection::sessionBus().call(message);
+    QDBusReply<QMap<QString, QVariantMap>> reply = QDBusConnection::sessionBus().call(message);
     if (reply.isValid()) {
-        const QMap<QString, QVariantMap> settingsMap = qvariant_cast<QMap<QString, QVariantMap>>(reply.value());
+        const QMap<QString, QVariantMap> settingsMap = reply.value();
         if (!settingsMap.isEmpty()) {
             const auto xdgColorSchemePref = static_cast<QXdgDesktopPortalThemePrivate::XdgColorschemePref>(settingsMap.value(appearanceInterface).value(colorSchemeKey).toUInt());
             d->colorScheme = QXdgDesktopPortalThemePrivate::colorSchemeFromXdgPref(xdgColorSchemePref);
             d->contrast = static_cast<Qt::ContrastPreference>(settingsMap.value(appearanceInterface).value(contrastKey).toUInt());
         }
+    } else {
+        qWarning() << "Call to org.freedesktop.portal.Settings.ReadAll failed" << reply.error();
     }
 
     QDBusConnection::sessionBus().connect(
